From 1d0ae279f07ec100597fc0e435436171e2f2ef11 Mon Sep 17 00:00:00 2001
From: fxcoudert <fxcoudert@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Sun, 15 Jun 2014 19:35:11 +0000
Subject: [PATCH 050/134] 	Backport from trunk 	PR fortran/45187

	* trans-decl.c (gfc_create_module_variable): Don't create
	Cray-pointee decls twice.

	* gfortran.dg/cray_pointers_10.f90: New file.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@211688 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/fortran/ChangeLog                          |  7 +++++++
 gcc/fortran/trans-decl.c                       |  4 ++--
 gcc/testsuite/ChangeLog                        |  6 ++++++
 gcc/testsuite/gfortran.dg/cray_pointers_10.f90 | 18 ++++++++++++++++++
 4 files changed, 33 insertions(+), 2 deletions(-)
 create mode 100644 gcc/testsuite/gfortran.dg/cray_pointers_10.f90

diff --git a/gcc/fortran/ChangeLog b/gcc/fortran/ChangeLog
index 4fbaf80e..8033c7a 100644
--- a/gcc/fortran/ChangeLog
+++ b/gcc/fortran/ChangeLog
@@ -1,3 +1,10 @@
+2014-06-15  Francois-Xavier Coudert  <fxcoudert@gcc.gnu.org>
+
+	Backport from trunk.
+	PR fortran/45187
+	* trans-decl.c (gfc_create_module_variable): Don't create
+	Cray-pointee decls twice.
+
 2014-05-26  Janne Blomqvist  <jb@gcc.gnu.org>
 
 	Backport from mainline
diff --git a/gcc/fortran/trans-decl.c b/gcc/fortran/trans-decl.c
index 7d235618..b4b0fc5 100644
--- a/gcc/fortran/trans-decl.c
+++ b/gcc/fortran/trans-decl.c
@@ -4084,8 +4084,8 @@ gfc_create_module_variable (gfc_symbol * sym)
     }
 
   /* Don't generate variables from other modules. Variables from
-     COMMONs will already have been generated.  */
-  if (sym->attr.use_assoc || sym->attr.in_common)
+     COMMONs and Cray pointees will already have been generated.  */
+  if (sym->attr.use_assoc || sym->attr.in_common || sym->attr.cray_pointee)
     return;
 
   /* Equivalenced variables arrive here after creation.  */
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index b58c46f..4fd2ef0 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,9 @@
+2014-06-15  Francois-Xavier Coudert  <fxcoudert@gcc.gnu.org>
+
+	Backport from trunk.
+	PR fortran/45187
+	* gfortran.dg/cray_pointers_10.f90: New file.
+
 2014-06-13  Peter Bergner  <bergner@vnet.ibm.com>
 
 	Backport from mainline
diff --git a/gcc/testsuite/gfortran.dg/cray_pointers_10.f90 b/gcc/testsuite/gfortran.dg/cray_pointers_10.f90
new file mode 100644
index 0000000..1ac98f3
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/cray_pointers_10.f90
@@ -0,0 +1,18 @@
+! { dg-do run }
+! { dg-options "-fcray-pointer" }
+!
+! PR fortran/45187
+!
+module foo
+  implicit none
+  real :: a
+  pointer(c_a, a)
+end module foo
+
+program test
+  use foo
+  real :: z
+  c_a = loc(z)
+  a = 42
+  if (z /= 42) call abort
+end program test
-- 
1.8.5.5

